lib/spack/spack/cmd/external.py

```diff
@@ -0,0 +1,204 @@
+
+    pkgs_dict = syaml.syaml_dict()
+    pkgs_dict[pkg_name] = pkg_dict
+
+    return pkgs_dict
+
+
+def external_find(args):
+    _get_external_packages(TestRepo())
+
+
+def _group_by_prefix(paths):
+    groups = defaultdict(set)
+    for p in paths:
+        groups[os.path.dirname(p)].add(p)
+    return groups.items()
+
+
+def _convert_to_iterable(single_val_or_multiple):
+    x = single_val_or_multiple
+    if x is None:
+        return []
+    elif isinstance(x, six.string_types):
+        return [x]
+    elif isinstance(x, spack.spec.Spec):
+        # Specs are iterable, but a single spec should be converted to a list
+        return [x]
+
+    try:
+        iter(x)
+        return x
+    except TypeError:
+        return [x]
+
+
+def _determine_base_dir(prefix):
+    assert os.path.isdir(prefix)
+    if os.path.basename(prefix) == 'bin':
+        return os.path.dirname(prefix)
+
+
+def _get_external_packages(repo, system_path_to_exe=None):
+    if not system_path_to_exe:
+        system_path_to_exe = _get_system_executables()
+
+    exe_pattern_to_pkgs = defaultdict(list)
+    for pkg in repo.all_packages():
+        if hasattr(pkg, 'executables'):
+            for exe in pkg.executables:
+                exe_pattern_to_pkgs[exe].append(pkg)
+
+    pkg_to_found_exes = defaultdict(set)
+    for path, exe in system_path_to_exe.items():
+        for exe_pattern, pkgs in exe_pattern_to_pkgs.items():
+            if re.search(exe_pattern, exe):
+                for pkg in pkgs:
+                    pkg_to_found_exes[pkg].add(path)
+
+    pkg_to_entries = defaultdict(list)
+    resolved_specs = {}  # spec -> exe found for the spec
+
+    # TODO: iterate through this in a predetermined order (e.g. by package
+    # name) to get repeatable results when there are conflicts.
+    for pkg, exes in pkg_to_found_exes.items():
+        for prefix, exes_in_prefix in _group_by_prefix(exes):
+            # TODO: multiple instances of a package can live in the same
+            # prefix, and a package implementation can return multiple specs
+            # for one prefix, but without additional details (e.g. about the
+            # naming scheme which differentiates them), the spec won't be
+            # usable.
+            if hasattr(pkg, 'determine_spec_details'):
+                specs = _convert_to_iterable(
+                    pkg.determine_spec_details(prefix, exes_in_prefix))
+            else:
+                # Note: packages which do not implement
+                # 'determine_spec_details' should use exact-match regular
+                # expressions
+                specs = [spack.spec.Spec(pkg.name)]
+
+            for spec in specs:
+                pkg_prefix = _determine_base_dir(prefix)
+
+                if not pkg_prefix:
+                    tty.debug("{0} does not end with a 'bin/' directory: it"
+                              " cannot be added as a Spack package"
+                              .format(prefix))
```

Comment:
Is this 100% true?  There are packages that call `env.prepend_path()` in `setup_run_environment()`.  Could we use that information to allow other directories?  e.g., if we see something that starts with the `prefix` in the `EnvModifications` for `PATH` after calling `setup_run_environment()`, we know that the package can have executables in that directory.  I don't think this is necessary for the first cut, but it's a way we could easily consider packages that don't exclusively use `bin`.

Reply:
Based on your suggestion, you are right that this package could strive to see whether it can make some use out of the prefix; generally (unless the package is exclusively provided externally) I think the `package.py` implementation would expect a typical Spack layout. There may be a couple packages that accommodate this now, and the `setup_run_environment` function could be updated to be more flexible.

IMO this could be added as a TODO and addressed later.

Reply:
You could easily add that later by adding match groups to the regex, and that’s more flexible anyway